sys/can/conn_can.c

```diff
@@ -0,0 +1,331 @@
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     can
+ * @{
+ *
+ * @file
+ * @brief       Implementation of generic can connection.
+ *              This file represents a generic can interface on top of Raw can,
+ *              ISOTP, TP 2.0 or BCM layers.
+ *
+ * @author      Aurelien Gonce <aurelien.gonce@altran.com>
+ * @author      Vincent Dupont <vincent@otakeys.com>
+ *
+ * @}
+ */
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <inttypes.h>
+#include <errno.h>
+
+#include "can/conn_can.h"
+#include "can/can.h"
+#include "can/raw.h"
+#include "thread.h"
+#ifdef MODULE_XTIMER
+#include "xtimer.h"
+#endif
+
+#define ENABLE_DEBUG (0)
+#include <debug.h>
+
+#define CONN_CAN_TIMEOUT        (3)
+#define _TIMEOUT_MSG_TYPE       (0x8000)
+#define _TIMEOUT_MSG_VALUE      (0xABCDEFAB)
+
+int conn_can_create(conn_can_t *conn, const void *info, size_t info_len,
+                    int ifnum, can_family_t family, int flags)
+{
+    if (conn == NULL) {
+        DEBUG("conn_can_create: conn object is NULL\n");
+        return -EINVAL;
+    }
+
+    int ret = 0;
+    kernel_pid_t my_pid = thread_getpid();
+
+    /* initialize conn structure */
+    conn->pid = my_pid;
+    conn->family = family;
+    conn->ifnum = ifnum;
+    conn->flags = flags;
+
+    switch (family) {
+#ifdef MODULE_CAN_RAW
+        case CONN_CAN_RAW:
+            ret = conn_can_set(conn, info, info_len, flags);
+            break;
+#endif
+
+        default:
+            DEBUG("conn_can_create: family not implemented\n");
+            (void) info;
+            (void) info_len;
+            ret = -EAFNOSUPPORT;
+            break;
+    }
+
+    return ret;
+}
+
+int conn_can_close(conn_can_t *conn)
+{
+    int ret = 0;
+
+    kernel_pid_t my_pid = conn->pid;
+    int ifnum = conn->ifnum;
+    can_family_t family = conn->family;
+
+    switch (family) {
+#ifdef MODULE_CAN_RAW
+        case CONN_CAN_RAW:
+            ret = raw_can_unsubscribe_rx_all(ifnum, my_pid);
+            break;
+#endif
+
+        default:
+            DEBUG("conn_can_close: Error: family not implemented\n");
+            (void) my_pid;
+            (void) ifnum;
+            ret = -EPFNOSUPPORT;
+            break;
+    }
+
+    conn->pid = 0;
+    conn->ifnum = CAN_IFNUM_UNDEF;
+    conn->family = CONN_CAN_UNDEF;
+
+    return ret;
+}
+
+int conn_can_recv(conn_can_t *conn, void *data, size_t max_len, uint32_t timeout)
+{
+    if (conn == NULL) {
+        DEBUG("conn_can_recv: conn object is NULL\n");
+        return -EINVAL;
+    }
+
+    can_family_t family = conn->family;
+
+#ifdef MODULE_XTIMER
+    xtimer_t timeout_timer;
+    msg_t timeout_msg;
+
+    if (timeout > 0) {
+        timeout_msg.type = _TIMEOUT_MSG_TYPE;
+        timeout_msg.content.value = _TIMEOUT_MSG_VALUE;
+        xtimer_set_msg(&timeout_timer, timeout, &timeout_msg, conn->pid);
```

Comment:
good point, but I wonder if it could make sense to allow creating/closing the connection from another thread than the one which receive. Then we could just remove the pid from the conn and simplify the code.